winml/lib/Api/LearningModelSession.cpp

```diff
@@ -198,54 +198,20 @@ LearningModelSession::EvaluateFeaturesAsync(
   return EvaluateAsync(binding, correlation_id);
 }
 
-static _winmla::IIOBinding*
-GetIOBinding(
-  winrt::com_ptr<winmlp::LearningModelBinding> binding_impl,
-  winml::LearningModel& model) {
-  // Get the IOBinding Collection, and bound outputs
-  com_ptr<_winmla::IIOBinding> io_binding;
-  io_binding.attach(binding_impl->BindingCollection());
-  auto& bound_output_names = io_binding->GetOutputNames();
-  std::unordered_set<std::string> bound_output_names_set(
-    bound_output_names.begin(),
-    bound_output_names.end());
-
-  // Get model output feature names
-  auto model_impl = model.as<winmlp::LearningModel>();
-  auto output_features = model_impl->OutputFeatures();
-  std::vector<ILearningModelFeatureDescriptor> output_descriptors(
-    begin(output_features),
-    end(output_features));
-
-  // Convert all output features to their feature names
-  std::vector<std::string> output_feature_names;
-  std::transform(
-    std::begin(output_descriptors),
-    std::end(output_descriptors),
-    std::back_inserter(output_feature_names),
-    [&](auto& descriptor) {
-      auto descriptor_native = descriptor.as<ILearningModelFeatureDescriptorNative>();
-      const wchar_t* p_name;
-      uint32_t size;
-      WINML_THROW_IF_FAILED(descriptor_native->GetName(&p_name, &size));
-      return WinML::Strings::UTF8FromUnicode(p_name, size);
-    });
-
-  // Find the set difference to determine if there are any unbound output features
-  std::vector<std::string> unbound_output_names;
-  std::copy_if(
-    std::begin(output_feature_names), std::end(output_feature_names),
-    std::inserter(unbound_output_names, std::begin(unbound_output_names)),
-    [&](const auto& outputFeatureName) {
-      return bound_output_names_set.find(outputFeatureName) == bound_output_names_set.end();
-    });
-
-  // Add all unbound outputs to the iobinding collection
-  for (const auto& unbound_output : unbound_output_names) {
-    WINML_THROW_IF_FAILED(io_binding->BindOutput(unbound_output, nullptr));
-  }
-
-  return io_binding.detach();
+// copied from onnxruntime_cxx_inline.h
+inline OrtStatus* OrtRun(
+    OrtSession * session, 
+    const Ort::RunOptions& run_options, 
+    const char* const* input_names, 
+    const Ort::Value* input_values, 
+    size_t input_count,
+    const char* const* output_names, 
+    Ort::Value* output_values, 
+    size_t output_count) {
+  static_assert(sizeof(Ort::Value) == sizeof(OrtValue*), "Value is really just an array of OrtValue* in memory, so we can reinterpret_cast safely");
+  auto ort_input_values = reinterpret_cast<const OrtValue**>(const_cast<Ort::Value*>(input_values));
```

Comment:
>reinterpret_cast [](start = 26, length = 16)

this means the windows.ai.machinelearning.dll also has RTTI on? It's off in the inbox build system. Do we really need it? is there another way to go back and forth between these two types?

Reply:
what's the question?    RTTI ?   are you thinking of changing this line to dynamic_cast?   we need reinterpret_cast here because there is zero lineage from Ort::Value to OrtValue.   the compiler does not know them to be related in anyway.    it's more a programmer choice that these are able to be casted this way.   check out line 251 in F:\src\github\microsoftonnxruntime_cxx_inline.h .   ryan : how about adding a comment to this part of the code showing where you copied it from (Ort::Session::Run in onnxruntime_cxx_inline.h).

---
In reply to: [351494357](https://github.com/microsoft/onnxruntime/pull/2504#discussion_r351494357) [](ancestors = 351494357)